doc/cpp_domdec.rst

```diff
@@ -0,0 +1,186 @@
+            node.num_gpus = hw::num_gpus()>0? 1: 0;
+
+    .. cpp:function:: node_info() = default
+
+        Default constructor (sets 1 CPU core and 0 GPUs).
+
+    .. cpp:function:: node_info(unsigned cores, unsigned gpus)
+
+        Constructor that sets the number of :cpp:var:`cores` and :cpp:var:`gpus` available.
+
+    .. cpp:member:: unsigned num_cpu_cores = 1
+
+        The number of CPU cores available.
+
+        By default it is assumed that there is one core available.
+
+    .. cpp:member:: unsigned num_gpus = 0
+
+        The number of GPUs available.
+
+        By default it is assumed that there are no GPUs.
+
+Load Balancers
+--------------
+
+Load balancing generates a :cpp:class:`domain_decomposition` given a :cpp:class:`recipe`
+and a description of the hardware on which the model will run. Currently Arbor provides
+one load balancer, :cpp:func:`partition_load_balance`, and more will be added over time.
+
+If the model is distributed with MPI, the partitioning algorithm for cells is
+distributed with MPI communication. The returned :cpp:class:`domain_decomposition`
+describes the cell groups on the local MPI rank.
+
+.. Note::
+    Power users know their model and the hardware resources that
+    they are running on very well, and know best how to partition their model
+    workload to get optimal performance.
+    The :cpp:class:`domain_decomposition` type is kept very simple, and is
+    independent of any load balancing algorithm, so power users can supply their
+    own domain decomposition without using one of the built-in load balancers.
+
+.. cpp:namespace:: arb
+
+.. cpp:function:: domain_decomposition partition_load_balance(const recipe& rec, hw::node_info nd)
+
+    Construct a :cpp:class:`domain_decomposition` that distributes the cells
+    in the model described by :cpp:var:`rec` over the hardware resources described
+    by `hw::node_info`.
+
+Decomposition
+-------------
+
+Documentation for the data structures used to describe domain decompositions.
+
+.. cpp:namespace:: arb
+
+.. cpp:enum-class:: backend_kind
+
+    Used to indicate which hardware backend to use for running a :cpp:class:`cell_group`.
+
+    .. cpp:enumerator:: multicore
+
+        Use multicore backend.
+
+    .. cpp:enumerator:: gpu
+
+        Use gpu back end.
+
+        .. Note::
+            Can only be set if the :cpp:class:`cell_group` type supports the GPU backend.
+
+.. cpp:class:: domain_decomposition
+
+    Describes a domain decomposition. Solely responsible for describing the
+    distribution of cells across cell groups and domains.
+    Holds cell group descriptions (:cpp:member:`groups`) for cells assigned to
+    the local domain, and a helper function (:cpp:member:`gid_domain`) used to
+    look up which domain a cell has been assigned to.
+    Also has meta-data about the number of cells in the global model, and the
+    number of domains over which the model is destributed.
```

Comment:
It's not clear if the object _specifies_ the number of cells and domains, or whether the values are derived elsewhere. As it's actually the latter, we should be clear in stating that these values must correspond to the actual number of cells in the recipe and the number of domains.

Ideally we wouldn't have the redundant specification of information, and we wouldn't need a copy in the domain decomposition object — possible future refactoring of this?

Reply:
I am actually quite comfortable with this description: it is just stating what the domain_decomposition represents. Where the values came from ins't really the concern of this description.

Reply:
I have added an extra "note" that lists requirements like `num_global_cells` must be the sum of `num_local_cells`.